extensions/PhabBugz/lib/Policy.pm

```diff
@@ -21,24 +19,30 @@ use List::Util qw(first);
 use Types::Standard -all;
 use Type::Utils;
 
-my $SearchResult = Dict[
-    phid      => Str,
-    type      => Str,
-    name      => Str,
-    shortName => Str,
-    fullName  => Str,
-    href      => Maybe[Str],
-    workflow  => Maybe[Str],
-    icon      => Str,
-    default   => Str,
-    rules     => ArrayRef[
+has 'phid'      => ( is => 'ro', isa => Str );
+has 'type'      => ( is => 'ro', isa => Str );
+has 'name'      => ( is => 'ro', isa => Str );
+has 'shortName' => ( is => 'ro', isa => Str );
+has 'fullName'  => ( is => 'ro', isa => Str );
+has 'href'      => ( is => 'ro', isa => Maybe[Str] );
+has 'workflow'  => ( is => 'ro', isa => Maybe[Str] );
+has 'icon'      => ( is => 'ro', isa => Str );
+has 'default'   => ( is => 'ro', isa => Str );
+has 'rules' => (
+    is  => 'ro',
+    isa => ArrayRef[
         Dict[
             action => Str,
             rule   => Str,
             value  => Maybe[ArrayRef[Str]]
         ]
     ]
-];
+);
+
+has 'rule_projects' => (
+    is => 'lazy',
+    isa => ArrayRef[Str],
+);
```

Comment:
For reasons I am not quite sure of right now, I could not get Revision.pm to work this way either. Something is different in how Moo pre-inits the variables and then when I get the data from Phabricator API, it cannot just simply copy the data struct into $self the way I do it in the current way. I would need to have to through each key/value pair in the API data and manually set the value for its corresponding Moo variable. Is there a way to do this more simply when loading the object with data? Using this code I get the following error right away:

bmo.feed_1          | [Fri Jan 26 20:07:02 2018] ERROR: Can't call method "rule_projects" on unblessed reference at /app/extensions/PhabBugz/lib/Feed.pm line 196.
bmo.feed_1          |  at /app/extensions/PhabBugz/lib/Feed.pm line 196.
bmo.feed_1          |   Bugzilla::Extension::PhabBugz::Feed::process_revision_change(Bugzilla::Extension::PhabBugz::Feed=HASH(0x5dd1228), "PHID-DREV-o674lixaeyxbfcr3e6oc", "conduit created D8: test.") called at /app/extensions/PhabBugz/lib/Feed.pm line 110
bmo.feed_1          |   Bugzilla::Extension::PhabBugz::Feed::feed_query(Bugzilla::Extension::PhabBugz::Feed=HASH(0x5dd1228)) called at /app/extensions/PhabBugz/lib/Feed.pm line 44
bmo.feed_1          |   eval {...} called at /app/extensions/PhabBugz/lib/Feed.pm line 42
bmo.feed_1          |   Bugzilla::Extension::PhabBugz::Feed::start(Bugzilla::Extension::PhabBugz::Feed=HASH(0x5dd1228)) called at /app/extensions/PhabBugz/lib/Daemon.pm line 97
bmo.feed_1          |   Bugzilla::Extension::PhabBugz::Daemon::gd_run(Bugzilla::Extension::PhabBugz::Daemon=HASH(0x1c72068)) called at local/lib/perl5/Daemon/Generic.pm line 192
bmo.feed_1          |   Daemon::Generic::new("Bugzilla::Extension::PhabBugz::Daemon") called at local/lib/perl5/Daemon/Generic.pm line 37
bmo.feed_1          |   Daemon::Generic::newdaemon() called at /app/extensions/PhabBugz/lib/Daemon.pm line 25
bmo.feed_1          |   Bugzilla::Extension::PhabBugz::Daemon::start("Bugzilla::Extension::PhabBugz::Daemon") called at extensions/PhabBugz/bin/phabbugz_feed.pl line 22

Reply:
Hmm. Nevermind my last comment. I see that you are doing new_from_query() which calls $class->new($result). This was not so easy in Revision.pm at the time as the API call has the data values nested under fields and not in the top level of the hash. So it became more code to do it with Moo that just using old-school accessors/setters.

In this case $result has the proper data from the API call, but  $class->new($result) is not returning a proper object to the caller. If I dump $current_policy in Feed.pm it is not blessed.